src/main/java/net/coderbot/iris/pipeline/ShadowRenderer.java

```diff
@@ -364,11 +364,119 @@ private Frustum createShadowFrustum() {
+
+		// Restore the old viewport
+		RenderSystem.viewport(0, 0, client.getWindow().getWidth(), client.getWindow().getHeight());
+	}
+
+	private void copyPreTranslucentDepth(LevelRendererAccessor levelRenderer) {
+		setProfilerTarget("translucent depth copy");
+
+		// Copy the content of the depth texture before rendering translucent content.
+		// This is needed for the shadowtex0 / shadowtex1 split.
+		RenderSystem.activeTexture(GL20C.GL_TEXTURE0);
+		RenderSystem.bindTexture(targets.getDepthTextureNoTranslucents().getTextureId());
+		GL20C.glCopyTexImage2D(GL20C.GL_TEXTURE_2D, 0, GL20C.GL_DEPTH_COMPONENT, 0, 0, resolution, resolution, 0);
+		RenderSystem.bindTexture(0);
+	}
+
+	private void renderEntities(LevelRendererAccessor levelRenderer, Frustum frustum, MultiBufferSource.BufferSource bufferSource, PoseStack modelView, double cameraX, double cameraY, double cameraZ, float tickDelta) {
+		EntityRenderDispatcher dispatcher = levelRenderer.getEntityRenderDispatcher();
+
+		int shadowEntities = 0;
+
+		levelRenderer.getLevel().getProfiler().push("cull");
+
+		List<Entity> renderedEntities = new ArrayList<>(32);
+
+		// TODO: I'm sure that this can be improved / optimized.
+		for (Entity entity : getLevel().entitiesForRendering()) {
+			if (!dispatcher.shouldRender(entity, frustum, cameraX, cameraY, cameraZ) || entity.isSpectator()) {
+				continue;
+			}
+
+			renderedEntities.add(entity);
+		}
+
+		setProfilerTarget("sort");
+
+		// Sort the entities by type first in order to allow vanilla's entity batching system to work better.
+		renderedEntities.sort(Comparator.comparingInt(entity -> entity.getType().hashCode()));
+
+		setProfilerTarget("build geometry");
+
+		for (Entity entity : renderedEntities) {
+			levelRenderer.invokeRenderEntity(entity, cameraX, cameraY, cameraZ, tickDelta, modelView, bufferSource);
+			shadowEntities++;
+		}
+
+		renderedShadowEntities = shadowEntities;
+	}
+
+	private void renderBlockEntities(LevelRendererAccessor levelRenderer, MultiBufferSource.BufferSource bufferSource, PoseStack modelView, double cameraX, double cameraY, double cameraZ, float tickDelta) {
+		levelRenderer.getLevel().getProfiler().push("build blockentities");
+
+		int shadowBlockEntities = 0;
+
+		// TODO: Use visibleChunks to cull block entities
+		for (BlockEntity entity : getLevel().blockEntityList) {
+			modelView.pushPose();
+			BlockPos pos = entity.getBlockPos();
+			modelView.translate(pos.getX() - cameraX, pos.getY() - cameraY, pos.getZ() - cameraZ);
+			BlockEntityRenderDispatcher.instance.render(entity, tickDelta, modelView, bufferSource);
+			modelView.popPose();
+
+			shadowBlockEntities++;
+		}
+
+		renderedShadowBlockEntities = shadowBlockEntities;
+	}
+
 	@Override
 	public void renderShadows(LevelRendererAccessor levelRenderer, Camera playerCamera) {
 		Minecraft client = Minecraft.getInstance();
 
-		levelRenderer.getLevel().getProfiler().popPush("shadows");
+		setProfilerTarget("shadows");
```

Comment:
I don't see the need for this particular change. It would make more sense to store the profiler at the head of the method or something to avoid repetition though,

Reply:
Will do

Reply:
done